### PR TITLE
utils: Fix segfaults caused by setargs()

### PR DESCRIPTION
**Pull request by gkarlos**
This PR fixes a number of segfaults caused when parsing options from a file ( `--opt-file`). Relevant issue #914 .

Note that there are still some leaks related to option parsing but they are not particularly tied to `setargs()`. They are caused by `exit()`-ing when an executable cannot be opened. This probably needs a new issue opened.

Signed-off-by: George Karlos gkarloscodes@gmail.com

### CODE REVIEW

File: cmds/record.c [hunk 1]
```diff
@@ -1554,6 +1554,9 @@ static void check_binary(struct opts *opts)
 		EM_X86_64, EM_ARM, EM_AARCH64, EM_386
 	};
 
+if ( !(opts->exename))
+	pr_err_ns("Cannot open '(null)' : Invalid filename.\n");
+
```
Comment (namhyung) [inline cmds/record.c]: Do you really need this?  It already checks the number of arguments and will print an error message if it misses exename.

File: uftrace.c [hunk 2]
```diff
@@ -912,7 +912,8 @@ static void parse_opt_file(int *argc, char ***argv, char *filename, struct opts
 		opts->idx = 0;
 	}
 
-	argp_parse(&file_argp, file_argc, file_argv,
+	if ( file_argv) /* workaround for when the file is empty */
+		argp_parse(&file_argp, file_argc, file_argv,
```
Comment (namhyung) [inline uftrace.c]: We don't add a space after open parenthesis.  Also you'd better use a pair of curly braces when the body of if statement is multi-line.  Please indent below lines as well.

File: uftrace.c [hunk 3]
```diff
@@ -1188,7 +1189,6 @@ int main(int argc, char *argv[])
 
 	if (opts.opt_file)
 		free_parsed_cmdline(argv - opts.idx);
-
```
Comment (namhyung) [inline uftrace.c]: Unnecessary change.

File: utils/utils.c [hunk 4]
```diff
@@ -712,56 +712,66 @@ char * json_quote(char *str, int *len)
...
+
 static int setargs(char *args, char **argv)
 {
 	int count = 0;
 
 	while (*args) {
-		/* ignore spaces */
-		if (isspace(*args)) {
+		/* ignore spaces and newlines */
+		if (isspace(*args) || isnewline(*args)) {
```
Comment (namhyung) [inline utils/utils.c]: AFAIK `isspace()` covers newlines too.

Comment (gkarlos) [reply utils/utils.c]: I believe you are right actually. I will revert that change

File: uftrace.c (hunk 2 above)
Comment (gkarlos) [reply uftrace.c]: This is also relevant when input is empty, i.e `""` or `"  "` or any number of spaces. So i guess i shall leave it out.

File: cmds/record.c (hunk 1 above)
Comment (namhyung) [reply cmds/record.c]: OK, then please write the exact command line to reproduce in the commit message.

File: utils/utils.c [hunk 5]
```diff
@@ -725,24 +728,28 @@ static int setargs(char *args, char **argv)
 
 		/* consider quotes and update argv */
 		if (*args == QUOTE) {
+			if (argv && *(++args) == QUOTE ) /* skip '' */
+				continue;
```
Comment (namhyung) [inline utils/utils.c]: Does it need to check `argv` here?  And I don't want to make the if statement complex.  Also I think the comment can be removed.  What about this?

```c
if (*args == QUOTE) {
   args++;
   if (*args == QUOTE)
      continue;
   ...
```

File: utils/utils.c [hunk 6]
```diff
@@ -725,24 +728,28 @@ static int setargs(char *args, char **argv)
...
+			while (*args && *args != DQUOTE) /* advance until " or end */
 				++args;
-			if (argv)
+			if (argv && *args)
 				*args = ' ';
 		}
 		else if (*args == '#') {
 			/* ignore comment line */
-			while (*args != '\n' || *args == '\0')
+			while (*args && !isnewline(*args) && (*args != '\0'))
```
Comment (namhyung) [inline utils/utils.c]: This can be `while (*args != '\n' && *args != '\0')`

**Comment by gkarlos** [comment]
@namhyung I think so. This fixes invalid filenames involving quotes that were causing segfaults before. e.g `echo "'" > opt` , `echo -n "'" > opt`, `echo "'''" > opt` etc.
